### ISSUE TITLE
TODO List

### ISSUE THREAD

**Issue by asarode** [florinpatrascu/neo4j_sips#2]
Thanks for making this! It's been really helpful so far. Do you have a roadmap or a todo list for this project? @ASCrookes said this in the other issue, I just wanted to make another issue to discuss it.

I'm still really new to Elixir but I'd love to help out any way I can!

**Comment by florinpatrascu** [comment]
Thank you too for the interest, I am glad you find this implementation useful. As I said in the other comment, I am too learning Elixir, therefore the more eyes seeing this code the better for all of its users :)

The TODO list is a big one, trust me, but here is the gist of it:
- introducing the **Model** concept. I started integrating @raw1z's Model design into mine, and I am very very close to release a first version. he did a wonderful job there, I wish I saw his project before starting mine. Very-very soon, you will be able to declare your neo4j_sips models as simple as this (exact from a current test suite):

``` elixir
defmodule Person do
  use Neo4j.Sips.Model

  field :name, required: true
  field :email, required: true, unique: true, format: ~r/\b[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}\b/
  field :age, type: :integer
  field :doe_family, type: :boolean, default: false # used for testing
  field :neo4j_sips, type: :boolean, default: true

  validate_with :check_age

  relationship :FRIEND_OF, Person
  relationship :MARRIED_TO, Person

  def check_age(model) do
    if model.age == nil || model.age <= 0 do
      {:age, "model.validation.invalid_age"}
    end
  end
end

```

the only obstacle holding me back from pushing this new release, is a model update bug I am currently working on, but there are only so many hours in a day
- clean up the tests and adding more tests for the Model implementation 
- consolidate the code and starting to work on the **Ecto** driver (my initial plan, when I started to work on this). Then Phoenix projects using neo4j_sips, like this one [Elixir, Phoenix - The Movies](https://github.com/florinpatrascu/movies_elixir_phoenix) would be a breeze to build :)
- implement a new Connection(-like) module to support [Bolt](https://dzone.com/articles/introducing-bolt-neo4js-upcoming-binary-protocol-p), Neo4j's upcoming binary protocol.

I was planning this week-end to push a development branch, containing the stable release of the Model code (incl tests. of course) and I'd appreciate if you guys would check it out. Sending pull requests is highly encouraged.

Thanks again for interest.

**Comment by asarode** [comment]
The model's definitely the biggest feature I was missing. I can test it out once the dev branch is pushed :)

**Comment by florinpatrascu** [comment]
dev branch available: https://github.com/florinpatrascu/neo4j_sips/tree/model_intro, have a look :)
Not ready for production yet.

**Comment by florinpatrascu** [comment]
Also, I started to remove the Mock support, from the tests. Was trying to maintain the model legacy tests as close as possible to their initial intent, but it's just more overhead that hinders on the new development. Rewriting most of the model tests where we used the Mock before. If you want to send any PR, please do not use Mock. Thank you! /cc @ASCrookes

**Comment by asarode** [comment]
I'll test it out tomorrow morning and let you know how it goes. Got a crazy project deadline tonight.

**Comment by florinpatrascu** [comment]
no worries

**Comment by florinpatrascu** [comment]
finally fixed the tests, they are all green now, see: the model_intro branch, for more details, if interested  /cc @ASCrookes. Will also open few new issues, for tracking the new development. 

Will close this issue, as the TODO list is available in branch now.

Cheers!
